### PR TITLE
Homework 3

### PR DESCRIPTION
**Pull request by iRezograf**
Refactoring done

### CODE REVIEW

File: src/test/java/ru/training/at/hw3/pageobjects/LoginPage.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        this.fieldPassword.sendKeys(password);
+    }
+
+    public void clickBtnEnter() {
+        this.btnEnter.click();
+    }
+
+    public String getUserName() {
+        WebDriverWait wait = new WebDriverWait(driver, 10);
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id='user-name']")));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/pageobjects/LoginPage.java]: To get rid of locator duplicating, please replace it with other expected condition
```suggestion
        wait.until(ExpectedConditions.visibilityOf(loggedUserName));
```

File: src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    @FindBy(xpath = "//*[@class='panel-body-list logs']/li[contains(string(),'Selen')]")
+    WebElement stringSelen;
+
+    @Getter
+    @FindBy(xpath = "//*[@class='panel-body-list logs']/li[contains(string(),'Wind')]")
+    WebElement stringWind;
+
+    @Getter
+    @FindBy(xpath = "//*[@class='panel-body-list logs']/li[contains(string(),'Water')]")
+    WebElement stringWater;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java]: Please replace these by one `List<WebElements>` that contains all log lines

File: src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        dropdownColors.click();
+    }
+
+    public void clickDropdownYellow() {
+        dropdownYellow.click();
+    }
+
+    public WebElement getLogList() {
+        return logList;
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java]: This method should return `List<String>` from a new field that contains all log lines

File: src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    WebElement benefitPractise;
+
+    @FindBy(xpath = "//span[contains(string(),'flexible and')]")
+    WebElement benefitFlexible;
+
+    @FindBy(xpath = "//span[contains(string(),'To be multiplatform')]")
+    WebElement benefitMultiplatform;
+
+    @FindBy(xpath = "//span[contains(string(),'Already have good base')]")
+    WebElement benefitBase;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java]: Please replace it with one `List<WebElement>` that contains all benefit texts
As a reference please use locator from `fourTextsOnTheIndexPageUnderIconsAndTheyHaveProperTextTest()` of hw2

File: src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        return this.benefitFlexible.getText();
+    }
+
+    public String getTextOfBenefitMultiplatform() {
+        return this.benefitMultiplatform.getText();
+    }
+
+    public String getTextOfBenefitBase() {
+        return this.benefitBase.getText();
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java]: These methods should be replaced by one returning `List<String>` from a new filed that contains all benefit texts

File: src/test/java/ru/training/at/hw3/pageobjects/BenefitIcon.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    private WebElement benefitPractice;
+
+    @FindBy(css = ".icons-benefit.icon-custom")
+    private WebElement benefitCustom;
+
+    @FindBy(css = ".icons-benefit.icon-multi")
+    private WebElement benefitMulti;
+
+    @FindBy(css = ".icons-benefit.icon-base")
+    private WebElement benefitBase;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/pageobjects/BenefitIcon.java]: You need only the count of icons, so please replace it with one `List<WebElement>` without defining of specific benefit types

File: src/test/java/ru/training/at/hw3/ExerciseOne.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    private void performLoginTest() {
+        loginPage.clickImgUser();
+        loginPage.inputFieldLogin(DataStore.getProperty("userName"));
+
+        loginPage.inputFieldPassword(DataStore.getProperty("password"));
+        loginPage.clickBtnEnter();
+        String actualUserNameAfterLogged = loginPage.getUserName();
+        softAssert.assertEquals(actualUserNameAfterLogged,
+                DataStore.getProperty("userNameAfterLogged"));
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/ExerciseOne.java]: Please parameterize methods in the following way
```java
public test() {
    openSiteByUrlTest(DataStore.getProperty("siteUrl"), DataStore.getProperty("browserTitle"));
    performLoginTest(DataStore.getProperty("userName"), DataStore.getProperty("password"), DataStore.getProperty("userNameAfterLogged"));
...
}

private void openSiteByUrlTest(final String siteUrl, final String expectedTitle) {
    driver.navigate().to(siteUrl);
    softAssert.assertEquals(driver.getTitle(), expectedTitle);
}

private void performLoginTest(String userName, String password, String expectedUserName) {
    loginPage.clickImgUser();
    loginPage.inputFieldLogin(userName);

    loginPage.inputFieldPassword(password);
    loginPage.clickBtnEnter();
    String actualUserNameAfterLogged = loginPage.getUserName();
    softAssert.assertEquals(actualUserNameAfterLogged, expectedUserName);
}

```

Other methods that are called from `@Test` method should be refactored in the same way

**Review by shoggy — CHANGES_REQUESTED** [review]

File: src/test/java/ru/training/at/hw3/pageobjects/LoginPage.java (hunk 1 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/LoginPage.java]: public String getUserName() {
        WebDriverWait wait = new WebDriverWait(driver, 10);
        wait.until(ExpectedConditions.visibilityOf(loggedUserName));
        String userName = loggedUserName.getText();
        return userName;
    }

File: src/test/java/ru/training/at/hw3/ExerciseOne.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    private LeftMenu leftMenu;
+    private SoftAssert softAssert;
+    private WebDriver driver;
+
+
+    @BeforeClass(groups = {"exercise_hw3.1"})
+    public void setUp() {
+        softAssert = new SoftAssert();
+        WebDriverManager.getInstance(CHROME).setup();
+        driver = new ChromeDriver();
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw3/ExerciseOne.java]: Let's move this to a separate class

File: src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java (hunk 3 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java:86]: Нет, она возвращает одну сконкатенированную строку, Точнее List c одной строкой.

File: src/test/java/ru/training/at/hw3/pageobjects/BenefitIcon.java (hunk 6 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/BenefitIcon.java]: @Getter
public class BenefitIcon {
    public WebDriver driver;

    @FindBy(css = ".benefit-icon")
    private List<WebElement> benefitIcons;

    public BenefitIcon(WebDriver driver) {
        PageFactory.initElements(driver, this);
        this.driver = driver;
    }
}

File: src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java (hunk 2 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java]: @FindBy(xpath = "//*[@class='panel-body-list logs']/li")
    List<WebElement> logList;

    public List<String> getLogListAsString() {
        List<String> list = new ArrayList<>();
        for (WebElement element : logList) {
            list.add(element.getText());
        }
        return list;
    }

File: src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java (hunk 4 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java]: @Getter
public class Benefits {
    public WebDriver driver;

    @FindBy(css = ".benefit-icon")
    private List<WebElement> benefitIcons;

    @FindBy(css = ".benefit-txt")
    private List<WebElement> benefitTxt;


    public Benefits(WebDriver driver) {
        PageFactory.initElements(driver, this);
        this.driver = driver;
    }

    public List<String> getBenefitTxtAsString() {
        List<String> list = new ArrayList<>();
        for (WebElement element : benefitTxt) {
            list.add(element.getText());
        }
        return list;
    }
}

File: src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java (hunk 5 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/BenefitText.java]: @Getter
public class Benefits {
    public WebDriver driver;

    @FindBy(css = ".benefit-icon")
    private List<WebElement> benefitIcons;

    @FindBy(css = ".benefit-txt")
    private List<WebElement> benefitTxt;


    public Benefits(WebDriver driver) {
        PageFactory.initElements(driver, this);
        this.driver = driver;
    }

    public List<String> getBenefitTxtAsString() {
        List<String> list = new ArrayList<>();
        for (WebElement element : benefitTxt) {
            list.add(element.getText());
        }
        return list;
    }
}

File: src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java (hunk 3 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/pageobjects/DifferentElementsPage.java]: @FindBy(xpath = "//*[@class='panel-body-list logs']/li")
List<WebElement> logList;

public List<String> getLogListAsString() {
    List<String> list = new ArrayList<>();
    for (WebElement element : logList) {
        list.add(element.getText());
    }
    return list;
}

File: src/test/java/ru/training/at/hw3/ExerciseOne.java (hunk 8 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/ExerciseOne.java]: package ru.training.at.hw3.util;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;

import static io.github.bonigarcia.wdm.config.DriverManagerType.CHROME;

public class DriverManager {
    WebDriver driver;

    public WebDriver setup() {
        WebDriverManager.getInstance(CHROME).setup();
        driver = new ChromeDriver();
        return driver;
    }
}

File: src/test/java/ru/training/at/hw3/ExerciseOne.java (hunk 7 above)
Comment (iRezograf) [reply src/test/java/ru/training/at/hw3/ExerciseOne.java]: @Test(groups = {"exercise_hw3.1"})
    public void allExerciseOneTest() {
        openSiteByUrlTest(
                DataStore.getProperty("siteUrl"),
                DataStore.getProperty("browserTitle"));

        performLoginTest(
                DataStore.getProperty("userName"),
                DataStore.getProperty("password"),
                DataStore.getProperty("userNameAfterLogged"));

        headerHaveProperTextsTest(
                getActualHeaderMenuItemHome(),
                getExpectedHeaderMenuItemHome());

        fourIconsInHomePageExistsTest(benefits.getBenefitIcons());

        fourTextsOnTheIndexPageUnderIconsAndTheyHaveProperTextTest(
                benefits.getBenefitTxtAsString(),
                getExpectedTextsOnTheIndexPageUnderIconsAndTheyHaveProperText());
        iframeWithFrameButtonExistTest(framePage);
        thereIsFrameButtonInTheIframeTest(framePage);
        switchToOriginalWindowBackTest(DataStore.getProperty("browserTitle"));
        thereAreFiveItemsInTheLeftSectionTest(
                leftMenu.getLeftMenuAsString(),
                getExceptedFiveItemsInTheLeftSection());
    }

**Review by shoggy — APPROVED** [review]
Good job!

**Pull request merged**